File: data/snippets/quiz013.cpp
class OrderProcessor {
public:
    // 检查库存函数
    // 这个函数用于检查产品是否有足够的库存
    // 参数:
    //   - productId: 产品ID
    //   - quantity: 需要的数量
    // 返回:
    //   - true: 如果库存足够
    //   - false: 如果库存不足
    bool checkInventory(int productId, int quantity) {
        // 获取当前库存数量
        int currentStock = getProductStock(productId);
        
        // 比较当前库存和需要的数量
        // 如果当前库存大于或等于需要的数量，返回true
        // 否则返回false
        if (currentStock >= quantity) {
            // 库存足够，返回true
            return true;
        } else {
            // 库存不足，返回false
            return false;
        }
    }
    
    // 处理订单函数
    // 这个函数处理客户订单
    // 它执行以下步骤:
    // 1. 检查库存
    // 2. 处理支付
    // 3. 发送确认邮件
    // 4. 更新库存
    bool processOrder(int orderId, int customerId, int productId, int quantity) {
        // 步骤1: 检查库存
        bool inStock = checkInventory(productId, quantity);
        
        // 如果库存不足，返回false
        if (!inStock) {
            // 记录库存不足错误
            // 库存不足，无法处理订单
            return false; // 返回处理失败
        }
        
        // 步骤2: 处理支付
        bool paymentSuccessful = processPayment(orderId, customerId);
        
        // 如果支付失败，返回false
        if (!paymentSuccessful) {
            // 支付失败，无法继续处理
            return false; // 返回处理失败
        }
        
        // 步骤3: 发送确认邮件
        sendConfirmationEmail(orderId, customerId);
        
        // 步骤4: 更新库存
        updateInventory(productId, quantity);
        
        // 所有步骤都成功完成，返回true
        return true; // 返回处理成功
    }
    
private:
    // 这个方法从数据库获取产品库存
    int getProductStock(int productId) {
        // 在实际应用中，这里会查询数据库
        // 现在只是返回一个假值
        return 100; // 假设所有产品库存都是100
    }
    
    // 这个方法处理客户支付
    bool processPayment(int orderId, int customerId) {
        // 在实际应用中，这里会调用支付网关
        // 现在只是返回一个假值
        return true; // 假设支付总是成功的
    }
    
    // 这个方法发送确认邮件
    void sendConfirmationEmail(int orderId, int customerId) {
        // 在实际应用中，这里会发送真实的邮件
        // 现在只是打印一条消息
        std::cout << "Confirmation email sent for order #" << orderId << std::endl;
    }
    
    // 这个方法更新库存
    void updateInventory(int productId, int quantity) {
        // 在实际应用中，这里会更新数据库
        // 现在只是打印一条消息
        std::cout << "Inventory updated for product #" << productId << std::endl;
    }
};

File: data/snippets/quiz013.java
public class OrderProcessor {
    
    // Check inventory function
    // This function is used to check if a product has sufficient inventory
    // Parameters:
    //   - productId: Product ID
    //   - quantity: Required quantity
    // Returns:
    //   - true: if inventory is sufficient
    //   - false: if inventory is insufficient
    public boolean checkInventory(int productId, int quantity) {
        // Get current inventory quantity
        int currentStock = getProductStock(productId);
        
        // Compare current inventory with required quantity
        // If current inventory is greater than or equal to required quantity, return true
        // Otherwise return false
        if (currentStock >= quantity) {
            // Inventory is sufficient, return true
            return true;
        } else {
            // Inventory is insufficient, return false
            return false;
        }
    }
    
    // Process order function
    // This function processes customer orders
    // It performs the following steps:
    // 1. Check inventory
    // 2. Process payment
    // 3. Send confirmation email
    // 4. Update inventory
    public boolean processOrder(int orderId, int customerId, int productId, int quantity) {
        // Step 1: Check inventory
        boolean inStock = checkInventory(productId, quantity);
        
        // If inventory is insufficient, return false
        if (!inStock) {
            // Log inventory shortage error
            // Inventory insufficient, cannot process order
            return false; // Return processing failure
        }
        
        // Step 2: Process payment
        boolean paymentSuccessful = processPayment(orderId, customerId);
        
        // If payment fails, return false
        if (!paymentSuccessful) {
            // Payment failed, cannot continue processing
            return false; // Return processing failure
        }
        
        // Step 3: Send confirmation email
        sendConfirmationEmail(orderId, customerId);
        
        // Step 4: Update inventory
        updateInventory(productId, quantity);
        
        // All steps completed successfully, return true
        return true; // Return processing success
    }
    
    // Calculate order total function
    // This function calculates the total amount for an order
    // It takes into account discounts and taxes
    public double calculateOrderTotal(int productId, int quantity, double discountPercentage) {
        // Get the product price from database
        double productPrice = getProductPrice(productId);
        
        // Calculate subtotal by multiplying price and quantity
        double subtotal = productPrice * quantity;
        
        // Calculate discount amount
        // Discount amount = subtotal * discount percentage
        double discountAmount = subtotal * (discountPercentage / 100.0);
        
        // Subtract discount from subtotal
        double afterDiscount = subtotal - discountAmount;
        
        // Calculate tax amount (assume 8% tax rate)
        double taxRate = 0.08; // 8% tax
        double taxAmount = afterDiscount * taxRate;
        
        // Add tax to get final total
        double finalTotal = afterDiscount + taxAmount;
        
        // Return the calculated total
        return finalTotal;
    }
    
    // This method gets product inventory from database
    private int getProductStock(int productId) {
        // In real application, this would query the database
        // For now, just return a dummy value
        return 100; // Assume all products have inventory of 100
    }
    
    // This method processes customer payment
    private boolean processPayment(int orderId, int customerId) {
        // In real application, this would call payment gateway
        // For now, just return a dummy value
        return true; // Assume payment is always successful
    }
    
    // This method sends confirmation email
    private void sendConfirmationEmail(int orderId, int customerId) {
        // In real application, this would send actual email
        // For now, just print a message
        System.out.println("Confirmation email sent for order #" + orderId);
    }
    
    // This method updates inventory
    private void updateInventory(int productId, int quantity) {
        // In real application, this would update the database
        // For now, just print a message
        System.out.println("Inventory updated for product #" + productId);
    }
    
    // This method gets product price from database
    private double getProductPrice(int productId) {
        // In real application, this would query the database
        // For now, just return a dummy value
        return 29.99; // Assume all products cost $29.99
    }
}
